docs/DeadmanAlert.md

```diff
@@ -0,0 +1,58 @@
+### Deadman Alert
+
+The `deadman` function is used in the context of a task to alert when a group stops reporting data.
+For example, say you have the following task definition:
+
+```
+option task = {
+    ...
+    every: 15m,
+    ...
+}
+from(bucket: "telegraf/autogen")
+    |> range(start: -1h)
+    |> filter(fn: (r) => r._measurement == "cpu")
+```
+
+In order to be alerted when the above task stops reporting data for a group, you would define a deadman alert like so:
+```
+option task = {
+    ...
+    every: 15m,
+    ...
+}
+f = (start, stop) =>
+    from(bucket: "telegraf/autogen")
+        |> range(start: start, stop: stop)
+        |> filter(fn: (r) => r._measurement == "cpu")
+
+deadman(f: f, period: 1h, offset: task.every)
+```
+
+where the deadman function is defined as follows:
+```
+deadman = (f, period, offset) => {
+
+    r = f(start: now() - offset - period, stop: now() - offset)
+        |> keys()
```

Comment:
What is the `keys()` function doing here? Can you provide trivial example tables for how this will detect missing series?

Reply:
I can provide an example, but the pattern:
```
... |> keys() |> drop(columns: ["_value"])
```
just reduces a stream down to its group key values (a table per group key). And then `diff` outputs the tables (group keys) that are in the first set, but not the second. And that gives you the series(groups) that stopped reporting.

Note I'm giving `diff` set semantics instead of bag semantics. I.e it performs an implicit deduplication of the rows. However, as part of this work, I think it would be best to add a transformation that deduplicates an entire stream. That way we can give `diff` bag semantics and perform deduplication before calling `diff`.

Reply:
I've added an example that hopefully clarifies things.

Reply:
@jlapacik I think I follow @nathanielc  here.   `keys` gets you the set of column names on a table, not the values.  

If you just want one row per series, then just do `limit(n:1)`

as mentioned earlier,  maybe: 
```
s = f(start: now() - period, stop: now())
        |> limit(n:1)
        |> map(fn: (r)  => ({r with _time  =  date.truncate(t: r._time, unit: 1h)}))
```
i wouldn't drop anything because the user may do that already.  or they want `_value` as part of the alert.    perhaps we need to add a parameter for what identifies a series?  and do  `keep(columns: seriesID)`

Reply:
I think you are looking for `keyValues`.

Reply:
Sorry there appears to be a lot of confusion over my use of `keys`. For each table that `keys` consumes, it outputs a single table consisting of the group key of the input table plus a `_value` column. By dropping the `_value` column you're just left with the group key (possibly duplicated over several rows).

@aanthony1243 thank you for the `limit` suggestion. I should use limit to reduce each output table to a single row representing the group key.

@affo you're correct in that `keyValues` would work as well. It functions similarly to `keys`. It just adds a `_key` and a `_value` column to the output table. These columns aren't really important. What we're after is the group key(series key).

@aanthony1243 to your point about dropping `_value`. `_value` is added by the `keys` function which is encapsulated within the deadman function. This is completely hidden from the user. What's important is the group/series keys. Deadman only cares when series stop reporting. In this context, only the group/series key is important. All other columns can and should be dropped so that we do a direct comparison of the series present in the first period that are not present in the second period.

Does this clarify why I'm using the `keys` function?

On a side note, I'd like to have a physical operation that retrieves just the series keys from storage.

Reply:
Oh, I see! Now I got it.

Reply:
I'd say that:

```
keep_group_cols = (tables=<-) => tables
    |> keys() // could even use `columns`, the point is to have a schema with only group key cols
    |> limit(n: 1)
    |> drop(columns: ["_value"])

r = f(start: now() - offset - period, stop: now() - offset) |> keep_group_cols()
s = f(start: now() - period, stop: now()) |> keep_group_cols()
```

Would help a lot

Reply:
if we generalize this concept, it's really saying "given this query, show me the signals that I heard in the previous time period that i haven't heard in the current time period".  

You are making influx-specific assumptions about what a `signal` is.  It may be safe to say, "signals are identified by the group key"  but this requires that the user has the data pre-grouped. Do we want to have the user tell the deadman what the signal-identifying columns are?